matching/src/main.ts

```diff
@@ -0,0 +1,308 @@
+import * as fs from "fs";
+import client from "./my_connection";
+import { Client } from "@elastic/elasticsearch";
+import { createDecompressor } from "lzma-native";
+import { split, map } from "event-stream";
+import { BiblObj, ElasticIndexHeader, Options, RawCrossrefData } from "./index";
+import { filterType, massage, round, sleep } from "./helpers";
+
+const red = "\x1b[31m";
+const reset = "\x1b[0m";
+
+const getHealth = async (client: Client): Promise<void> => {
+  const health = await client.cluster.health();
+  console.log("ES Health --", health);
+};
+
+const indexData = async (options: Options, client: Client) => {
+  const { body: indexExists } = await client.indices
+    .exists({
+      index: options.indexName,
+    })
+    .catch((err) => {
+      console.error("Index exist check failed");
+      throw err;
+    });
+
+  if (indexExists) {
+    // remove previous index if it exists
+    await client.indices.delete({ index: options.indexName }).catch((err) => {
+      console.error("Index deletion failed");
+      throw err;
+    });
+    console.log(`Index ${options.indexName} have been deleted`);
+  }
+
+  await client.indices
+    .create({ index: options.indexName, body: analyzers })
+    .catch((err) => {
+      console.error("Index deletion failed");
+      throw err;
+    });
+  console.log(`Index ${options.indexName} created`);
+
+  await client.indices
+    .putMapping({
+      index: options.indexName,
+      type: options.docType,
+      body: mapping,
+    })
+    .catch((err) => {
+      console.error("Mapping failed");
+      throw err;
+    });
+  console.log(`Mapping created for index ${options.indexName}`);
+
+  await index(options, client);
+};
+
+async function processAction(options: Options) {
+  if (options.action === "health") {
+    getHealth(client);
+  } else if (options.action === "index") {
+    indexData(options, client);
+  }
+}
+
+const buildBibliographicField = (obj: Partial<BiblObj>): string => {
+  return [
+    obj.first_author,
+    obj.title,
+    obj.journal,
+    obj.abbreviated_journal,
+    obj.volume,
+    obj.issue,
+    obj.first_page,
+    obj.year,
+  ]
+    .filter(Boolean)
+    .join(" ");
+};
+
+/**
+ * This function removes some non-used stuff from a crossref work entry,
+ * in particular the citation information, which represent a considerable
+ * amount of data.
+ */
+const createBiblObj = (data: RawCrossrefData): BiblObj => {
+  // Just keep the fields we want to index
+  // - Main fields (in the mapping)
+  const obj: Partial<BiblObj> = {
+    title: data.title,
+    DOI: data.DOI,
+    journal: data["container-title"],
+    abbreviated_journal: data["short-container-title"],
+    volume: data.volume,
+    type: data.type,
+    year: getYear(data),
+  };
+
+  if (data.author?.length > 0) {
+    const { authors, firstAuthorName } = data.author.reduce(
+      (acc, author) => {
+        if (author.sequence === "first") {
```

Comment:
I asked crossref support about this, because this 'sequence' attribute is not available in the last dump, maybe consider the first one in the list as first too, also the language attribut is not available (maybe it is not relevant for you..)

Reply:
I wasn't aware that the sequence attribute was missing to be honest and maybe the order should be considered as well. Although I'm not sure that that would always yield the correct result. Do you have any more insight on this?

Reply:
to build my actual index , I have considered either the sequence attribute or  the order in the list, I think in most cases the list order represents the real order..